File: haskell/euler1_6.hs
-- Euler1 in Haskell

import Debug.Trace

debug = flip trace

slice :: [Int] -> Int -> Int -> [Int]
slice xs from to = do
    let result = drop from $ take to xs `debug` ("slice: " ++ show xs ++ ", " ++ show from ++ ", " ++ show to)
    result                              `debug` ("sliced: " ++ show result)

euler :: [Int] -> Int
euler xs =
    if length xs == 0
        then 0 `debug` ("returning 0")
        else do
            let pivot = (length xs) `div` 2
            let pre  = euler [(head xs)..pivot]
            let post = euler (slice xs (pivot+1) (length xs))
            let pivotval = xs !! pivot
            if pivotval `mod` 3 == 0 || pivotval `mod` 5 == 0
                then pre + pivotval + post
                else pre + post

euler1 :: Int -> Int
euler1 n = euler [1..n]

main :: IO ()
main = do
    let result = euler1 999 
    putStrLn $ "euler1 = " ++ show result

File: f#/euler1_6.fs
(* Euler1 in F# *)

let rec range i j = i :: if i=j then [] else (range (i+1) j));;

let euler n =
    let rec f xs =
		if xs = [] then 0;

        let pivot = length xs / 2;
		let pre = euler (range 0 pivot);
		let post = euler (range (pivot+1) (length xs);
		let pivotval = xs[pivot];
		if pivotval % 3 = 0 || pivotval % 5 = 0 then
			pre + pivotval + post;
		pre + post;
	in f range 0 n;;

Printf.printf "Euler1 = %d\n" (euler 999);;

